File: app/src/main/java/com/zofers/zofers/create/CreateOfferFirstFragment.java
package com.zofers.zofers.create;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.RadioGroup;
import android.widget.Spinner;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.google.android.material.snackbar.Snackbar;
import com.google.android.material.textfield.TextInputLayout;
import com.zofers.zofers.R;
import com.zofers.zofers.model.Offer;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Locale;


public class CreateOfferFirstFragment extends CreateOfferBaseFragment {

    private Spinner countrySpinner;
    private EditText cityEdittext;
    private RadioGroup radioGroup;
    //    private RadioButton expensesMeRB;
    private EditText costEdittext;
    private EditText currencyEdittext;
    private TextInputLayout cityTIL;
    private TextInputLayout costTIL;
    private TextInputLayout currencyTIL;
    private View countryUnderline;
    View view;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
         view = inflater.inflate(R.layout.fragment_create_offer_first, container, false);

        countrySpinner = view.findViewById(R.id.country_spinner);
        ArrayAdapter<String> adapter = new ArrayAdapter<String>(getContext(), android.R.layout.simple_list_item_1, countryList());
        countrySpinner.setAdapter(adapter);

        cityEdittext = view.findViewById(R.id.city_editText);
        costEdittext = view.findViewById(R.id.cost_editText);
        currencyEdittext = view.findViewById(R.id.currency_editText);
        cityTIL = view.findViewById(R.id.city_TIL);
        costTIL = view.findViewById(R.id.cost_TIL);
        currencyTIL = view.findViewById(R.id.currency_TIL);
        countryUnderline = view.findViewById(R.id.country_underline_view);

        radioGroup = view.findViewById(R.id.radioGroup);
        radioGroup.setOnCheckedChangeListener(new RadioGroup.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(RadioGroup group, int checkedId) {
                boolean freeForGuest = checkedId == R.id.expensesMe;
                costEdittext.setEnabled(!freeForGuest);
                currencyEdittext.setEnabled(!freeForGuest);
                if (freeForGuest) {
                    costEdittext.setText("");
                }
            }

        });
        return view;
    }

    @Override
    public int getProgress() {
        return 33;
    }

    @Override
    public boolean validFilled() {
        boolean validFilled = true;
        if (countrySpinner.getSelectedItemPosition() == 0) {
//            Snackbar.make(view,"Please select country", Snackbar.LENGTH_SHORT).show();
            countryUnderline.setBackgroundColor(getContext().getResources().getColor(R.color.error_color));
            validFilled = false;
        }
        if (cityEdittext.getText().toString().trim().length() == 0) {
            cityTIL.setError(" ");
            validFilled = false;
        }
        if (radioGroup.getCheckedRadioButtonId() == -1 && validFilled) {
            Snackbar.make(view,"Please select who will pay Expenses", Snackbar.LENGTH_SHORT).show();
            validFilled = false;
        }
        if (radioGroup.getCheckedRadioButtonId() != R.id.expensesMe && costEdittext.getText().length() == 0) {
            costTIL.setError(" ");
            validFilled = false;
        }
        if (currencyEdittext.getText().toString().trim().length() == 0) {
            currencyTIL.setError(" ");
            validFilled = false;
        }
        return validFilled;
    }

    @Override
    public CreateOfferBaseFragment nextFragment() {
        return new CreateOfferSecoundFragment();
    }

    @Override
    public Offer fillOffer(@NonNull Offer offer) {
        offer.setCountry(countrySpinner.getSelectedItem().toString());
        offer.setCity(cityEdittext.getText().toString().trim());
        offer.setCostMode(costMode());
        if (radioGroup.getCheckedRadioButtonId() != R.id.expensesMe) {
            String cost = costEdittext.getText().toString();
            offer.setCost(Integer.parseInt(cost));
            String currency = currencyEdittext.getText().toString().trim();
            offer.setCurrency(currency);
        }
        return offer;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
    }

    private ArrayList<String> countryList() {
        Locale[] locale = Locale.getAvailableLocales();
        ArrayList<String> countries = new ArrayList<String>();
        String country;
        for (Locale loc : locale) {
            country = loc.getDisplayCountry();
            if (country.length() > 0 && !countries.contains(country)) {
                countries.add(country);
            }
        }
        Collections.sort(countries, String.CASE_INSENSITIVE_ORDER);
        countries.add(0, "Select Country");
        return countries;
    }

    private int costMode() {
        switch (radioGroup.getCheckedRadioButtonId()) {
            case R.id.expensesMe:
                return Offer.COST_MODE_CREATOR;
            case R.id.expensesBoth:
                return Offer.COST_MODE_BOTH;
            case R.id.expensesGuest:
                return Offer.COST_MODE_GUEST;
            default:
                return Offer.COST_MODE_CREATOR;
        }
    }
}

File: app/src/main/java/com/zofers/zofers/ui/create/CreateOfferFirstFragment.kt
package com.zofers.zofers.ui.create

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.core.widget.doOnTextChanged
import com.google.android.material.snackbar.Snackbar
import com.google.android.material.textfield.TextInputLayout
import com.zofers.zofers.R
import com.zofers.zofers.model.Offer
import java.util.*

class CreateOfferFirstFragment : CreateOfferBaseFragment() {
	private var countrySpinner: Spinner? = null
	private var countryText: TextView? = null
	private var cityEdittext: EditText? = null
	private var radioGroup: RadioGroup? = null
	//    private RadioButton expensesMeRB;
	private var costEdittext: EditText? = null
	private var costLayout: ViewGroup? = null
	private var currencyEdittext: EditText? = null
	private var cityTIL: TextInputLayout? = null
	private var costTIL: TextInputLayout? = null
	private var currencyTIL: TextInputLayout? = null
	private var countryUnderline: View? = null
	private lateinit var root: View

	override val progress: Int
		get() = 33

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
		root = inflater.inflate(R.layout.fragment_create_offer_first, container, false)
		countrySpinner = root.findViewById(R.id.country_spinner)
		countryText = root.findViewById(R.id.country_text)
		val adapter = ArrayAdapter(context!!, android.R.layout.simple_list_item_1, countryList())
		countrySpinner?.adapter = adapter
		cityEdittext = root.findViewById(R.id.city_editText)
		costEdittext = root.findViewById(R.id.cost_editText)
		costLayout = root.findViewById(R.id.cost_layout)
		currencyEdittext = root.findViewById(R.id.currency_editText)
		cityTIL = root.findViewById(R.id.city_TIL)
		costTIL = root.findViewById(R.id.cost_TIL)
		currencyTIL = root.findViewById(R.id.currency_TIL)
		countryUnderline = root.findViewById(R.id.country_underline_view)
		radioGroup = root.findViewById(R.id.radioGroup)
		radioGroup?.setOnCheckedChangeListener { _, checkedId ->
			val freeForGuest = checkedId == R.id.expensesMe
			costEdittext?.isEnabled = !freeForGuest
			currencyEdittext?.isEnabled = !freeForGuest
			costLayout?.visibility = if (freeForGuest) View.GONE else View.VISIBLE
			if (freeForGuest) {
				costEdittext?.setText("")
			}
		}
		cityEdittext?.doOnTextChanged { _, _, _, _ ->
			cityTIL?.error = null
		}
		costEdittext?.doOnTextChanged { _, _, _, _ ->
			 costTIL?.error = null
		}
		currencyEdittext?.doOnTextChanged { _, _, _, _ ->
			currencyTIL?.error = null
		}
		countrySpinner?.onItemSelectedListener = object : AdapterView.OnItemSelectedListener {
			override fun onNothingSelected(parent: AdapterView<*>?) {}
			override fun onItemSelected(parent: AdapterView<*>?, view: View?, position: Int, id: Long) {
				countryUnderline?.setBackgroundColor(context!!.resources.getColor(R.color.gray_transparent))
				countryText?.setTextColor(context!!.resources.getColor(R.color.gray_transparent))
			}

		}
		return root
	}

	override fun validFilled(): Boolean {
		var validFilled = true
		if (countrySpinner?.selectedItemPosition == 0) { //            Snackbar.make(view,"Please select country", Snackbar.LENGTH_SHORT).show();
			countryUnderline?.setBackgroundColor(context!!.resources.getColor(R.color.error_color))
			countryText?.setTextColor(context!!.resources.getColor(R.color.error_color))
			validFilled = false
		}
		if (cityEdittext?.text.toString().trim { it <= ' ' }.isEmpty()) {
			cityTIL?.error = " "
			validFilled = false
		}
		if (radioGroup?.checkedRadioButtonId == -1 && validFilled) {
			Snackbar.make(root, "Please select who will pay Expenses", Snackbar.LENGTH_SHORT).show()
			validFilled = false
		}
		if (radioGroup?.checkedRadioButtonId != R.id.expensesMe && costEdittext!!.text.length == 0) {
			costTIL?.error = " "
			validFilled = false
		}
		if (currencyEdittext!!.text.toString().trim { it <= ' ' }.isEmpty()) {
			currencyTIL?.error = " "
			validFilled = false
		}
		return validFilled
	}

	override fun nextFragment(): CreateOfferBaseFragment {
		return CreateOfferSecoundFragment()
	}

	override fun fillFields(offer: Offer) {
		countrySpinner?.setSelection(countryList().indexOf(offer.country))
		cityEdittext?.setText(offer.city)
		radioGroup?.check(
				when (offer.costMode) {
					Offer.COST_MODE_BOTH -> R.id.expensesBoth
					Offer.COST_MODE_GUEST -> R.id.expensesGuest
					else -> R.id.expensesMe
				}
		)
		costEdittext?.setText(offer.cost.toString())
		currencyEdittext?.setText(offer.currency)
	}

	override fun fillOffer(offer: Offer): Offer {
		offer.country = countrySpinner?.selectedItem.toString()
		offer.countryCode = getCountryCode(offer.country)
		offer.city = cityEdittext?.text.toString().trim { it <= ' ' }.toLowerCase(Locale.ROOT)
		offer.costMode = costMode()
		if (radioGroup?.checkedRadioButtonId != R.id.expensesMe) {
			val cost = costEdittext?.text.toString()
			offer.cost = cost.toInt()
			offer.currency = currencyEdittext?.text.toString().trim { it <= ' ' }
		}
		return offer
	}

	private fun countryList(): ArrayList<String> {
		val locale = Locale.getAvailableLocales()
		val countries = ArrayList<String>()
		var country: String
		for (loc in locale) {
			country = loc.displayCountry
			if (country.isNotEmpty() && !countries.contains(country)) {
				countries.add(country)
			}
		}
		Collections.sort(countries, java.lang.String.CASE_INSENSITIVE_ORDER)
		countries.add(0, getString(R.string.select_country))
		return countries
	}

	private fun getCountryCode(country: String): String {
		Locale.getAvailableLocales().forEach { loc ->
			if (loc.displayCountry == country) {
				return loc.country
			}
		}
		return ""
	}

	private fun costMode(): Int {
		return when (radioGroup!!.checkedRadioButtonId) {
			R.id.expensesMe -> Offer.COST_MODE_CREATOR
			R.id.expensesBoth -> Offer.COST_MODE_BOTH
			R.id.expensesGuest -> Offer.COST_MODE_GUEST
			else -> Offer.COST_MODE_CREATOR
		}
	}
}